### PR TITLE
fix issues in discovering ruff in pip build environments

### PR DESCRIPTION
**Pull request by jsurany**
## Summary

Changes in this PR https://github.com/astral-sh/ruff/pull/13591 did not allow correct discovery in pip build environments.

```python
# both of these variables are tuple[str, str] (length is 2)
first, second = os.path.split(paths[0]), os.path.split(paths[1])

# so these length checks are guaranteed to fail even for build environment folders
if (
    len(first) >= 3
    and len(second) >= 3 
    ...
)
```

~~Here we instead use `pathlib`, and we check all `pip-build-env-` paths for the folder that is expected to contain the `ruff` executable.~~

Here we update the logic to more properly split out the path components that we use for `pip-build-env-` inspection.

## Test Plan

I've checked this manually against a workflow that was failing, I'm not sure what to do for real tests. The same issues apply as with the previous PR.

### CODE REVIEW

File: python/ruff/__main__.py [hunk 1]
```diff
@@ -35,24 +36,20 @@ def find_ruff_bin() -> str:
...
     #
+    # Expect to find ruff in <prefix>/pip-build-env-<rand>/overlay/bin/ruff
+    #
     # See: https://github.com/pypa/pip/blob/102d8187a1f5a4cd5de7a549fd8a9af34e89a54f/src/pip/_internal/build_env.py#L87
-    paths = os.environ.get("PATH", "").split(os.pathsep)
-    if len(paths) >= 2:
-        first, second = os.path.split(paths[0]), os.path.split(paths[1])
-        # Search for both an `overlay` and `normal` folder within a `pip-build-env-{random}` folder. (The final segment
-        # of the path is the `bin` directory.)
+    for path_str in os.environ.get("PATH", "").split(os.pathsep):
```
Comment (gaborbernat) [inline python/ruff/__main__.py:42]: We should only check the last path, not all.

File: python/ruff/__main__.py [hunk 2]
```diff
@@ -35,24 +36,20 @@ def find_ruff_bin() -> str:
...
         if (
-            len(first) >= 3
-            and len(second) >= 3
-            and first[-3].startswith("pip-build-env-")
-            and first[-2] == "overlay"
-            and second[-3].startswith("pip-build-env-")
-            and second[-2] == "normal"
+            len(path_parts) >= 3
+            and path_parts[-3].startswith("pip-build-env-")
+            and path_parts[-2] == "overlay"
```
Comment (gaborbernat) [inline python/ruff/__main__.py:48]: You removed also asserting the `normal` entry.

File: python/ruff/__main__.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import os
 import sys
 import sysconfig
+from pathlib import Path
```
Comment (gaborbernat) [inline python/ruff/__main__.py:4]: `pathlib` is expensive to import during the startup, so we actively wanted to avoid using it; use os.path instead

**Review by gaborbernat — COMMENTED** [review]
Needs some work IMHO.

File: python/ruff/__main__.py [hunk 4]
```diff
@@ -35,24 +49,22 @@ def find_ruff_bin() -> str:
...
-        # of the path is the `bin` directory.)
-        if (
-            len(first) >= 3
-            and len(second) >= 3
-            and first[-3].startswith("pip-build-env-")
-            and first[-2] == "overlay"
-            and second[-3].startswith("pip-build-env-")
-            and second[-2] == "normal"
-        ):
-            # The overlay must contain the ruff binary.
```
Comment (gaborbernat) [inline python/ruff/__main__.py:52]: SHould put back this comment.

File: python/ruff/__main__.py [hunk 5]
```diff
@@ -35,24 +49,22 @@ def find_ruff_bin() -> str:
 
     # Search for pip-specific build environments.
     #
+    # Expect to find ruff in <prefix>/pip-build-env-<rand>/overlay/bin/ruff
```
Comment (gaborbernat) [inline python/ruff/__main__.py:52]: also theNormal part

File: python/ruff/__main__.py [hunk 6]
```diff
@@ -3,6 +3,20 @@
 import sysconfig
 
 
+def get_path_parts(path: str) -> list[str]:
```
Comment (gaborbernat) [inline python/ruff/__main__.py:6]: we only do checks against the last 3 elements, so perhaps we should stop after 3?

Comment (jsurany) [reply python/ruff/__main__.py]: I'm ok with stopping at 3, not a fan of returning a tuple.

**Review by gaborbernat — COMMENTED** [review]
👍

File: python/ruff/__main__.py [hunk 7]
```diff
@@ -35,24 +52,25 @@ def find_ruff_bin() -> str:
...
-            and first[-3].startswith("pip-build-env-")
-            and first[-2] == "overlay"
-            and second[-3].startswith("pip-build-env-")
-            and second[-2] == "normal"
-        ):
-            # The overlay must contain the ruff binary.
-            candidate = os.path.join(first, ruff_exe)
-            if os.path.isfile(candidate):
-                return candidate
+    maybe_overlay = get_last_three_path_parts(paths[0])
```
Comment (charliermarsh) [inline python/ruff/__main__.py:60]: Can't this lead to an index-out-of-bounds error?

**Review by charliermarsh — COMMENTED** [review]
What was the actual problem here?

**Comment by gaborbernat** [comment]
> What was the actual problem here?

https://github.com/astral-sh/ruff/pull/13591 does not work 😆 this fixes that.

File: python/ruff/__main__.py (hunk 7 above)
Comment (gaborbernat) [reply python/ruff/__main__.py:60]: Yeah there is an if statement missing here, ` if len(paths) >= 2:` that got removed by mistake.

Comment (jsurany) [reply python/ruff/__main__.py]: good catch, updated.

File: python/ruff/__main__.py [hunk 8]
```diff
@@ -35,22 +52,24 @@ def find_ruff_bin() -> str:
...
+        maybe_normal = get_last_three_path_parts(paths[1])
         if (
-            len(first) >= 3
-            and len(second) >= 3
-            and first[-3].startswith("pip-build-env-")
-            and first[-2] == "overlay"
-            and second[-3].startswith("pip-build-env-")
-            and second[-2] == "normal"
+            len(maybe_normal) >= 3
+            and maybe_normal[-1].startswith("pip-build-env-")
```
Comment (charliermarsh) [inline python/ruff/__main__.py:65]: Isn't this wrong? The path is like `<prefix>/pip-build-env-<rand>/overlay/bin/ruff`. So then `maybe_normal` would be `['ruff', 'bin', 'overlay']`, right?

Comment (jsurany) [reply python/ruff/__main__.py:65]: The paths from `$PATH` don't include the actual executable. For instance for `maybe_overlay` the path we parse would look like `<prefix>/pip-build-env-<rand>/overlay/bin`, so `maybe_overlay` is `["bin", "overlay", "pip-build-env-<rand>"]`

File: python/ruff/__main__.py [hunk 9]
```diff
@@ -35,22 +35,40 @@ def find_ruff_bin() -> str:
...
     if len(paths) >= 2:
-        first, second = os.path.split(paths[0]), os.path.split(paths[1])
-        # Search for both an `overlay` and `normal` folder within a `pip-build-env-{random}` folder. (The final segment
-        # of the path is the `bin` directory.)
+
+        def get_last_three_path_parts(path: str) -> list[str]:
+            """Return a list of up to the last three parts of a path."""
+            parts = []
+
+            while len(parts) < 3:
```
Comment (charliermarsh) [inline python/ruff/__main__.py:49]: I changed this to just use `len(parts)` instead of a separate index, which I believe is equivalent.

**Pull request merged**